rust/otap-dataflow/crates/otap/src/pdata.rs

```diff
@@ -129,85 +137,283 @@ impl OtlpProtoBytes {
+    /// points, log records).
+    #[must_use]
+    pub fn num_items(&self) -> usize {
+        match self {
+            Self::OtlpBytes(value) => value.num_items(),
+            Self::OtapArrowRecords(value) => value.num_items(),
         }
     }
 }
 
-/* -------- Helper trait implementations -------- */
+/* -------- Trait implementations -------- */
 
 /// Helper methods that internal representations of OTAP PData should implement
-trait OtapPdataHelpers {
+pub trait OtapPayloadHelpers {
     /// Returns the type of signal represented by this `OtapPdata` instance.
     fn signal_type(&self) -> SignalType;
+
+    /// Number of items.
+    fn num_items(&self) -> usize;
+
+    /// Return true if there is no data.
+    fn is_empty(&self) -> bool;
+
+    /// Takes the payload, leaving an empty payload behind.
+    fn payload(&mut self) -> Self;
 }
 
-impl OtapPdataHelpers for OtlpProtoBytes {
+impl OtapPayloadHelpers for OtapArrowRecords {
+    fn signal_type(&self) -> SignalType {
+        match self {
+            Self::Logs(_) => SignalType::Logs,
+            Self::Metrics(_) => SignalType::Metrics,
+            Self::Traces(_) => SignalType::Traces,
+        }
+    }
+
+    fn payload(&mut self) -> Self {
+        match self {
+            Self::Logs(value) => Self::Logs(std::mem::take(value)),
+            Self::Metrics(value) => Self::Metrics(std::mem::take(value)),
+            Self::Traces(value) => Self::Traces(std::mem::take(value)),
+        }
+    }
+
+    fn is_empty(&self) -> bool {
+        match self {
+            Self::Logs(_) => {
+                self.get(otel_arrow_rust::proto::opentelemetry::arrow::v1::ArrowPayloadType::Logs)
+                    .is_none_or(|batch| batch.num_rows() == 0)
+            }
+            Self::Traces(_) => {
+                self.get(otel_arrow_rust::proto::opentelemetry::arrow::v1::ArrowPayloadType::Spans)
+                    .is_none_or(|batch| batch.num_rows() == 0)
+            }
+            Self::Metrics(_) => {
+                self.get(otel_arrow_rust::proto::opentelemetry::arrow::v1::ArrowPayloadType::UnivariateMetrics)
+                    .is_none_or(|batch| batch.num_rows() == 0)
+            }
+        }
+    }
+
+    fn num_items(&self) -> usize {
+        match self {
+            Self::Logs(_) => {
+                self
+                    .get(otel_arrow_rust::proto::opentelemetry::arrow::v1::ArrowPayloadType::Logs)
+                    .map_or(0, |batch| batch.num_rows())
+            }
+            Self::Traces(_) => {
+		self
+                    .get(otel_arrow_rust::proto::opentelemetry::arrow::v1::ArrowPayloadType::Spans)
+                    .map_or(0, |batch| batch.num_rows())
+            }
+            Self::Metrics(_) => {
+                self.get(otel_arrow_rust::proto::opentelemetry::arrow::v1::ArrowPayloadType::UnivariateMetrics)
+                    .map_or(0, |batch| batch.num_rows())
+            }
+        }
+    }
+}
```

Comment:
Should this just call `self.batch_length()`?
https://github.com/open-telemetry/otel-arrow/blob/1baa5f5be1526aaf2077ba0b58ff4ce94c4755f4/rust/otel-arrow-rust/src/otap.rs#L96-L102

Note - `batch_length` for metrics counts the datapoints, not the rows in the `UnivariateMetrics` `RecordBatch`:
https://github.com/open-telemetry/otel-arrow/blob/1baa5f5be1526aaf2077ba0b58ff4ce94c4755f4/rust/otel-arrow-rust/src/otap.rs#L368-L387